telegram/_messageentity.py

```diff
@@ -140,6 +140,92 @@ def de_json(cls, data: Optional[JSONDict], bot: "Bot") -> Optional["MessageEntit
+        Unicode and the `utf_16_entities` are calculated in UTF-16.
+
+        ```python
+        text = "𠌕 bold 𝄢 italic underlined: 𝛙𝌢𑁍"
+        #         ^^^^   ^^^^^^             ^^^
+        unicode_entities = [
+            MessageEntity(offset=2, length=4, type=MessageEntity.BOLD),
+            MessageEntity(offset=9, length=6, type=MessageEntity.ITALIC),
+            MessageEntity(offset=28, length=3, type=MessageEntity.UNDERLINE),
+        ]
+        utf_16_entities = MessageEntity.adjust_message_entities_to_utf_16(
+            text, unicode_entities
+        )
+        await bot.send_message(
+            chat_id=123,
+            text=text,
+            entities=utf_16_entities,
+        )
+        # utf_16_entities[0]: offset=3, length=4
+        # utf_16_entities[1]: offset=11, length=6
+        # utf_16_entities[2]: offset=30, length=6
+        ```
+
+        Args:
+            text (:obj:`str`): The text that the entities belong to
+            entities (List[:class:`telegram.MessageEntity`]): Sequence of entities
+                with offset and length calculated in Unicode
+
+        Returns:
+            List[:class:`telegram.MessageEntity`]: Sequence of entities with offset and length
+                calculated in UTF-16 encoding
+        """
+        offsets_set = set()
+        # collect all offsets occured
+        for entity in entities:
+            offsets_set.add(entity.offset)
+            offsets_set.add(entity.offset + entity.length)
+        # get sorted offsets
+        offsets_list = sorted(offsets_set)
+        # indexes_map stores the index of each `cut_offset` in `offsets_list`
+        indexes_map = {cut_offset: i for i, cut_offset in enumerate(offsets_list)}
+        # calculate the length of each slice in utf-16
+        accumulated_length = 0
+        utf16_lengths = []
+        # what utf16_lengths stores:
+        # utf16_lengths[indexes_map[offset]] = len(text[:offset].encode("utf-16-le")) // 2
+        for i, cut_offset in enumerate(offsets_list):
+            last_cut_offset = offsets_list[i - 1] if i > 0 else 0
+            text_slice = text[last_cut_offset:cut_offset]
+            accumulated_length += len(text_slice.encode("utf-16-le")) // 2
+            utf16_lengths.append(accumulated_length)
+        # get the final output entites
+        out = []
+        for entity in entities:
+            translated_offset = utf16_lengths[indexes_map[entity.offset]]
+            translated_length = (
+                utf16_lengths[indexes_map[entity.offset + entity.length]] - translated_offset
+            )
+            new_entity = MessageEntity(
+                type=entity.type,
+                offset=translated_offset,
+                length=translated_length,
+                url=entity.url,
+                user=entity.user,
+                language=entity.language,
+                custom_emoji_id=entity.custom_emoji_id,
+                api_kwargs={**entity.api_kwargs},
+            )
```

Comment:
new_entity = copy.copy(entity)

```suggestion
            new_entity = copy.copy(entity)
            with new_entity._unfreeze():
                new_entity.offset = translated_offset
                new_entity.length = translated_length
```

`._unfreeze()` is an internal helper that allows us to override attributes after initialization if required.
Donig it this way prevents that we forget adding new arguments if they are added to `MessageEntity` :)